{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import qiime2\n",
    "from qiime2.plugins import (\n",
    "    cutadapt, demux, dada2, feature_table, metadata,\n",
    "    greengenes2, taxa, feature_classifier,\n",
    "    vsearch\n",
    ")\n",
    "\n",
    "from qiime2 import Artifact, Metadata\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "\n",
    "path = {\n",
    "    \"art\" : \"../data/artifacts/mock/\",\n",
    "    \"vis\" : \"../visualizations/mock/\",\n",
    "    \"res\" : \"../data/resources/\"\n",
    " }\n",
    "\n",
    "for filepath in path.values():\n",
    "    os.makedirs(filepath, exist_ok=True)\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    complement = {'A': 'T', 'C': 'G',\n",
    "                  'G': 'C', 'T': 'A',\n",
    "                  'N': 'N',\n",
    "                  'R': 'Y', 'Y': 'R'}\n",
    "    return \"\".join(complement[base] for base in reversed(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create mock metadata\n",
    "metadata_df = pd.DataFrame({\n",
    "    'sampleid': ['2476', '2477'], 'mock' : [0, 1]}).set_index('sampleid')\n",
    "\n",
    "metadata_q2 = qiime2.Metadata(metadata_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Quality control \n",
    "## 1.1. Adapter trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import raw data with manifest\n",
    "raw_seqs = qiime2.Artifact.import_data('SampleData[PairedEndSequencesWithQuality]',\n",
    "                                       '../data/manifest-mock.tsv', view_type='PairedEndFastqManifestPhred33V2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The length of ITS region is variable, compared to 16S. 16S rRNA is long, and MiSeq will not cover the whole region with a single read. During ITS sequencing, full ITS region get sequenced and sometimes adapter sequences and primers got into resulting reads. We need to clean them, as they can \"mask\" as real biological DNA, while they are not. Knowing used primers and adapters we will remove them with `cutadapt`.  \n",
    "Primers and adapter sequences for MiSeq:\n",
    "- forward: TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGTAAAAGTCGTAACAAGGTTTC\n",
    "- reverse: GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTTCAAAGAYTCGATGATTCAC\n",
    "\n",
    "- Primers should be passed to `cutadapt` without overhang - [discussion on QIIME2 Forum](https://forum.qiime2.org/t/remove-primer-in-paired-end-demultiplexed-file/17376/12)\n",
    "- Primers were used in another [paper](https://www.biorxiv.org/content/10.1101/2021.07.19.452952v1.full)\n",
    "- Reverse complements should also be passed as in [tutorial](https://forum.qiime2.org/t/fungal-its-analysis-tutorial/7351)\n",
    "```\n",
    "--p-front-f GTAAAAGTCGTAACAAGGTTTC \\\n",
    "--p-front-r GTTCAAAGAYTCGATGATTCAC \\\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The commands to be run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: cutadapt --cores 8 --error-rate 0.1 --times 1 --overlap 3 --minimum-length 1 -q 0,0 --quality-base 33 -o /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2476_0_L001_R1_001.fastq.gz -p /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2476_2_L001_R2_001.fastq.gz --adapter GTGAATCATCGARTCTTTGAAC --front GTAAAAGTCGTAACAAGGTTTC -A GAAACCTTGTTACGACTTTTAC -G GTTCAAAGAYTCGATGATTCAC /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2476_0_L001_R1_001.fastq.gz /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2476_2_L001_R2_001.fastq.gz\n",
      "\n",
      "This is cutadapt 4.6 with Python 3.8.15\n",
      "Command line parameters: --cores 8 --error-rate 0.1 --times 1 --overlap 3 --minimum-length 1 -q 0,0 --quality-base 33 -o /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2476_0_L001_R1_001.fastq.gz -p /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2476_2_L001_R2_001.fastq.gz --adapter GTGAATCATCGARTCTTTGAAC --front GTAAAAGTCGTAACAAGGTTTC -A GAAACCTTGTTACGACTTTTAC -G GTTCAAAGAYTCGATGATTCAC /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2476_0_L001_R1_001.fastq.gz /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2476_2_L001_R2_001.fastq.gz\n",
      "Processing paired-end reads on 8 cores ...\n",
      "Finished in 1.200 s (9.867 Âµs/read; 6.08 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:            121,623\n",
      "  Read 1 with adapter:                   1,945 (1.6%)\n",
      "  Read 2 with adapter:                     404 (0.3%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Pairs that were too short:                   0 (0.0%)\n",
      "Pairs written (passing filters):       121,623 (100.0%)\n",
      "\n",
      "Total basepairs processed:    73,217,046 bp\n",
      "  Read 1:    36,608,523 bp\n",
      "  Read 2:    36,608,523 bp\n",
      "Quality-trimmed:                       0 bp (0.0%)\n",
      "  Read 1:             0 bp\n",
      "  Read 2:             0 bp\n",
      "Total written (filtered):     73,188,024 bp (100.0%)\n",
      "  Read 1:    36,585,962 bp\n",
      "  Read 2:    36,602,062 bp\n",
      "\n",
      "=== First read: Adapter 1 ===\n",
      "\n",
      "Sequence: GTGAATCATCGARTCTTTGAAC; Type: regular 3'; Length: 22; Trimmed: 1731 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 5.0%\n",
      "  C: 8.4%\n",
      "  G: 18.5%\n",
      "  T: 68.1%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t481\t1900.4\t0\t481\n",
      "4\t432\t475.1\t0\t432\n",
      "5\t746\t118.8\t0\t746\n",
      "6\t2\t29.7\t0\t2\n",
      "27\t6\t0.0\t2\t5 1\n",
      "29\t2\t0.0\t2\t2\n",
      "33\t1\t0.0\t2\t0 1\n",
      "48\t1\t0.0\t2\t1\n",
      "49\t1\t0.0\t2\t1\n",
      "61\t2\t0.0\t2\t1 1\n",
      "62\t1\t0.0\t2\t1\n",
      "65\t9\t0.0\t2\t9\n",
      "79\t2\t0.0\t2\t2\n",
      "80\t2\t0.0\t2\t1 1\n",
      "83\t1\t0.0\t2\t1\n",
      "164\t1\t0.0\t2\t1\n",
      "179\t1\t0.0\t2\t1\n",
      "180\t9\t0.0\t2\t8 0 1\n",
      "182\t1\t0.0\t2\t1\n",
      "195\t4\t0.0\t2\t4\n",
      "223\t1\t0.0\t2\t1\n",
      "226\t3\t0.0\t2\t3\n",
      "232\t1\t0.0\t2\t1\n",
      "245\t3\t0.0\t2\t3\n",
      "250\t10\t0.0\t2\t9 1\n",
      "252\t1\t0.0\t2\t1\n",
      "255\t1\t0.0\t2\t1\n",
      "262\t1\t0.0\t2\t1\n",
      "264\t1\t0.0\t2\t1\n",
      "272\t1\t0.0\t2\t1\n",
      "276\t3\t0.0\t2\t3\n",
      "\n",
      "\n",
      "=== First read: Adapter 2 ===\n",
      "\n",
      "Sequence: GTAAAAGTCGTAACAAGGTTTC; Type: regular 5'; Length: 22; Trimmed: 214 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "19\t1\t0.0\t1\t1\n",
      "21\t3\t0.0\t2\t1 2\n",
      "22\t210\t0.0\t2\t187 21 2\n",
      "\n",
      "\n",
      "=== Second read: Adapter 3 ===\n",
      "\n",
      "Sequence: GAAACCTTGTTACGACTTTTAC; Type: regular 3'; Length: 22; Trimmed: 110 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 40.0%\n",
      "  C: 22.7%\n",
      "  G: 15.5%\n",
      "  T: 21.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t90\t1900.4\t0\t90\n",
      "4\t2\t475.1\t0\t2\n",
      "5\t1\t118.8\t0\t1\n",
      "6\t4\t29.7\t0\t4\n",
      "171\t1\t0.0\t2\t0 0 1\n",
      "176\t1\t0.0\t2\t1\n",
      "180\t1\t0.0\t2\t0 1\n",
      "208\t1\t0.0\t2\t0 0 1\n",
      "222\t1\t0.0\t2\t0 1\n",
      "252\t1\t0.0\t2\t1\n",
      "262\t1\t0.0\t2\t1\n",
      "264\t1\t0.0\t2\t1\n",
      "274\t1\t0.0\t2\t1\n",
      "276\t1\t0.0\t2\t0 1\n",
      "277\t2\t0.0\t2\t2\n",
      "278\t1\t0.0\t2\t1\n",
      "\n",
      "\n",
      "=== Second read: Adapter 4 ===\n",
      "\n",
      "Sequence: GTTCAAAGAYTCGATGATTCAC; Type: regular 5'; Length: 22; Trimmed: 294 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t176\t1900.4\t0\t176\n",
      "4\t3\t475.1\t0\t3\n",
      "5\t1\t118.8\t0\t1\n",
      "7\t1\t7.4\t0\t1\n",
      "21\t2\t0.0\t2\t0 1 1\n",
      "22\t110\t0.0\t2\t63 27 20\n",
      "23\t1\t0.0\t2\t0 0 1\n",
      "\n",
      "Command: cutadapt --cores 8 --error-rate 0.1 --times 1 --overlap 3 --minimum-length 1 -q 0,0 --quality-base 33 -o /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2477_1_L001_R1_001.fastq.gz -p /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2477_3_L001_R2_001.fastq.gz --adapter GTGAATCATCGARTCTTTGAAC --front GTAAAAGTCGTAACAAGGTTTC -A GAAACCTTGTTACGACTTTTAC -G GTTCAAAGAYTCGATGATTCAC /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2477_1_L001_R1_001.fastq.gz /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2477_3_L001_R2_001.fastq.gz\n",
      "\n",
      "This is cutadapt 4.6 with Python 3.8.15\n",
      "Command line parameters: --cores 8 --error-rate 0.1 --times 1 --overlap 3 --minimum-length 1 -q 0,0 --quality-base 33 -o /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2477_1_L001_R1_001.fastq.gz -p /tmp/q2-CasavaOneEightSingleLanePerSampleDirFmt-fq52rinq/2477_3_L001_R2_001.fastq.gz --adapter GTGAATCATCGARTCTTTGAAC --front GTAAAAGTCGTAACAAGGTTTC -A GAAACCTTGTTACGACTTTTAC -G GTTCAAAGAYTCGATGATTCAC /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2477_1_L001_R1_001.fastq.gz /tmp/qiime2/vbezshapkin/data/5ba0a8e6-26dd-4ea9-8fba-9b6073d1068e/data/2477_3_L001_R2_001.fastq.gz\n",
      "Processing paired-end reads on 8 cores ...\n",
      "Finished in 0.709 s (14.441 Âµs/read; 4.15 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total read pairs processed:             49,075\n",
      "  Read 1 with adapter:                  48,767 (99.4%)\n",
      "  Read 2 with adapter:                  46,587 (94.9%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Pairs that were too short:                   0 (0.0%)\n",
      "Pairs written (passing filters):        49,075 (100.0%)\n",
      "\n",
      "Total basepairs processed:    29,543,150 bp\n",
      "  Read 1:    14,771,575 bp\n",
      "  Read 2:    14,771,575 bp\n",
      "Quality-trimmed:                       0 bp (0.0%)\n",
      "  Read 1:             0 bp\n",
      "  Read 2:             0 bp\n",
      "Total written (filtered):     26,930,506 bp (91.2%)\n",
      "  Read 1:    13,368,883 bp\n",
      "  Read 2:    13,561,623 bp\n",
      "\n",
      "=== First read: Adapter 1 ===\n",
      "\n",
      "Sequence: GTGAATCATCGARTCTTTGAAC; Type: regular 3'; Length: 22; Trimmed: 13161 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 92.5%\n",
      "  C: 6.7%\n",
      "  G: 0.3%\n",
      "  T: 0.5%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by 'A' extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "    Ignore this warning when trimming primers.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t5\t766.8\t0\t5\n",
      "4\t2\t191.7\t0\t2\n",
      "5\t1\t47.9\t0\t1\n",
      "25\t15\t0.0\t2\t15\n",
      "27\t2\t0.0\t2\t2\n",
      "29\t3\t0.0\t2\t3\n",
      "32\t1\t0.0\t2\t1\n",
      "33\t1\t0.0\t2\t0 0 1\n",
      "44\t2\t0.0\t2\t2\n",
      "45\t119\t0.0\t2\t88 28 3\n",
      "46\t12021\t0.0\t2\t11858 157 6\n",
      "47\t763\t0.0\t2\t599 156 8\n",
      "48\t102\t0.0\t2\t64 33 5\n",
      "49\t16\t0.0\t2\t15 1\n",
      "50\t11\t0.0\t2\t6 3 2\n",
      "51\t1\t0.0\t2\t1\n",
      "52\t3\t0.0\t2\t3\n",
      "55\t1\t0.0\t2\t1\n",
      "61\t8\t0.0\t2\t8\n",
      "65\t3\t0.0\t2\t3\n",
      "85\t1\t0.0\t2\t1\n",
      "112\t4\t0.0\t2\t4\n",
      "122\t1\t0.0\t2\t1\n",
      "123\t1\t0.0\t2\t1\n",
      "135\t1\t0.0\t2\t1\n",
      "148\t2\t0.0\t2\t2\n",
      "172\t2\t0.0\t2\t2\n",
      "179\t1\t0.0\t2\t0 1\n",
      "180\t7\t0.0\t2\t7\n",
      "183\t1\t0.0\t2\t1\n",
      "185\t1\t0.0\t2\t1\n",
      "209\t2\t0.0\t2\t2\n",
      "226\t7\t0.0\t2\t7\n",
      "228\t1\t0.0\t2\t1\n",
      "248\t1\t0.0\t2\t1\n",
      "250\t12\t0.0\t2\t12\n",
      "252\t1\t0.0\t2\t1\n",
      "254\t10\t0.0\t2\t10\n",
      "255\t4\t0.0\t2\t4\n",
      "257\t1\t0.0\t2\t1\n",
      "263\t1\t0.0\t2\t1\n",
      "264\t1\t0.0\t2\t1\n",
      "268\t1\t0.0\t2\t1\n",
      "269\t1\t0.0\t2\t1\n",
      "271\t2\t0.0\t2\t2\n",
      "273\t3\t0.0\t2\t3\n",
      "275\t1\t0.0\t2\t1\n",
      "277\t6\t0.0\t2\t6\n",
      "278\t2\t0.0\t2\t2\n",
      "279\t2\t0.0\t2\t2\n",
      "\n",
      "\n",
      "=== First read: Adapter 2 ===\n",
      "\n",
      "Sequence: GTAAAAGTCGTAACAAGGTTTC; Type: regular 5'; Length: 22; Trimmed: 35606 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "16\t4\t0.0\t1\t3 1\n",
      "17\t2\t0.0\t1\t0 2\n",
      "19\t1\t0.0\t1\t1\n",
      "20\t89\t0.0\t2\t1 42 46\n",
      "21\t1361\t0.0\t2\t487 789 85\n",
      "22\t33918\t0.0\t2\t32200 1604 114\n",
      "23\t229\t0.0\t2\t52 171 6\n",
      "158\t2\t0.0\t2\t1 1\n",
      "\n",
      "\n",
      "=== Second read: Adapter 3 ===\n",
      "\n",
      "Sequence: GAAACCTTGTTACGACTTTTAC; Type: regular 3'; Length: 22; Trimmed: 7933 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 1.3%\n",
      "  C: 1.0%\n",
      "  G: 96.6%\n",
      "  T: 1.1%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by 'G' extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "    Ignore this warning when trimming primers.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t110\t766.8\t0\t110\n",
      "4\t53\t191.7\t0\t53\n",
      "6\t1\t12.0\t0\t1\n",
      "45\t84\t0.0\t2\t49 30 5\n",
      "46\t7221\t0.0\t2\t6956 237 28\n",
      "47\t389\t0.0\t2\t248 133 8\n",
      "48\t47\t0.0\t2\t23 15 9\n",
      "49\t3\t0.0\t2\t1 2\n",
      "50\t3\t0.0\t2\t1 1 1\n",
      "51\t2\t0.0\t2\t0 0 2\n",
      "52\t3\t0.0\t2\t3\n",
      "58\t4\t0.0\t2\t2 1 1\n",
      "65\t1\t0.0\t2\t1\n",
      "180\t1\t0.0\t2\t0 1\n",
      "218\t1\t0.0\t2\t1\n",
      "226\t1\t0.0\t2\t1\n",
      "245\t1\t0.0\t2\t1\n",
      "250\t3\t0.0\t2\t3\n",
      "254\t2\t0.0\t2\t2\n",
      "255\t1\t0.0\t2\t1\n",
      "267\t1\t0.0\t2\t0 1\n",
      "277\t1\t0.0\t2\t1\n",
      "\n",
      "\n",
      "=== Second read: Adapter 4 ===\n",
      "\n",
      "Sequence: GTTCAAAGAYTCGATGATTCAC; Type: regular 5'; Length: 22; Trimmed: 38654 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20-22 bp: 2\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t15\t766.8\t0\t15\n",
      "4\t2\t191.7\t0\t2\n",
      "5\t1\t47.9\t0\t1\n",
      "18\t5\t0.0\t1\t3 2\n",
      "19\t6\t0.0\t1\t3 1 2\n",
      "20\t51\t0.0\t2\t6 17 28\n",
      "21\t972\t0.0\t2\t223 634 115\n",
      "22\t37399\t0.0\t2\t33933 2871 595\n",
      "23\t203\t0.0\t2\t49 132 22\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n"
     ]
    }
   ],
   "source": [
    "forward_primer = \"GTAAAAGTCGTAACAAGGTTTC\"\n",
    "reverse_primer = \"GTTCAAAGAYTCGATGATTCAC\"\n",
    "\n",
    "trimmed = cutadapt.methods.trim_paired(\n",
    "    demultiplexed_sequences=raw_seqs,\n",
    "    # reverse complement of reverse primer\n",
    "    adapter_f = [reverse_complement(reverse_primer)],\n",
    "    front_f = [forward_primer],\n",
    "    # reverse complement of forward primer\n",
    "    adapter_r = [reverse_complement(forward_primer)],\n",
    "    front_r = [reverse_primer],\n",
    "    cores = 8\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../visualizations/mock/quality-plot-fungi.qzv'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "quality_vis = demux.visualizers.summarize(trimmed.trimmed_sequences, n = 100_000)\n",
    "quality_vis.visualization.save(path[\"vis\"] + \"quality-plot-fungi.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2. DADA2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada.R --input_directory /tmp/tmp0xn0virg/forward --input_directory_reverse /tmp/tmp0xn0virg/reverse --output_path /tmp/tmp0xn0virg/output.tsv.biom --output_track /tmp/tmp0xn0virg/track.tsv --filtered_directory /tmp/tmp0xn0virg/filt_f --filtered_directory_reverse /tmp/tmp0xn0virg/filt_r --truncation_length 240 --truncation_length_reverse 169 --trim_left 0 --trim_left_reverse 0 --max_expected_errors 2.0 --max_expected_errors_reverse 2.0 --truncation_quality_score 2 --min_overlap 12 --pooling_method independent --chimera_method consensus --min_parental_fold 4 --allow_one_off False --num_threads 32 --learn_min_reads 1000000\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "package âoptparseâ was built under R version 4.2.3 \n",
      "Loading required package: Rcpp\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 4.2.2 (2022-10-31) \n",
      "DADA2: 1.26.0 / Rcpp: 1.0.12 / RcppParallel: 5.1.6 \n",
      "2) Filtering ..\n",
      "3) Learning Error Rates\n",
      "19731600 total bases in 82215 reads from 2 samples will be used for learning the error rates.\n",
      "13894335 total bases in 82215 reads from 2 samples will be used for learning the error rates.\n",
      "3) Denoise samples ..\n",
      "..\n",
      "5) Remove chimeras (method = consensus)\n",
      "6) Report read numbers through the pipeline\n",
      "7) Write output\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'../visualizations/mock/feature-table-fungi.qzv'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_reads = dada2.methods.denoise_paired(\n",
    "    trimmed.trimmed_sequences,\n",
    "    trunc_len_f=240, trunc_len_r=169, n_threads=32,\n",
    "    min_fold_parent_over_abundance=4\n",
    ")\n",
    "\n",
    "qc_reads.denoising_stats.save(path[\"art\"] + \"denoise-stats-fungi.qza\")\n",
    "qc_reads.table.save(path[\"art\"] + \"feature-table-fungi.qza\")\n",
    "qc_reads.representative_sequences.save(path[\"art\"] + \"rep-seqs-fungi.qza\")\n",
    "\n",
    "\n",
    "metadata.visualizers.tabulate(input=qc_reads.denoising_stats.view(Metadata)).visualization.save(path[\"vis\"] + \"denoise-stats-fungi.qzv\")\n",
    "feature_table.visualizers.summarize(qc_reads.table).visualization.save(path[\"vis\"] + \"feature-table-fungi.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Taxonomical classification - ITS1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Run out of memory, reset tmpdir, runs bashes script internally, doesn't respect fish shell :/\n",
    "# # classification command\n",
    "# export TMPDIR=./tmp/\n",
    "# qiime feature-classifier classify-sklearn \\\n",
    "#     --i-reads data/artifacts/mock/rep-seqs-fungi.qza \\\n",
    "#     --i-classifier data/resources/unite_ver9_99_25.07.2023-Q2-2023.9.qza \\\n",
    "#     ---p-n-jobs 8 \\\n",
    "#     --o-classification data/artifacts/mock/rep-seqs-unite-its.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load classification data\n",
    "tax = Artifact.load(path[\"art\"] + \"rep-seqs-unite-its.qza\")\n",
    "table = Artifact.load(path[\"art\"] + \"feature-table-fungi.qza\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../visualizations/mock/taxonomy-barplot-fungi.qzv'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vis = taxa.visualizers.barplot(\n",
    "    table=table,\n",
    "    taxonomy=tax,\n",
    "    metadata=metadata_q2\n",
    ")\n",
    "\n",
    "vis.visualization.save(path[\"vis\"] + \"taxonomy-barplot-fungi.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-2024.2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
